book/src/design/circuit/gadgets/sinsemilla.md

```diff
@@ -0,0 +1,145 @@
+  \vdots    & \vdots     &   \vdots  & \vdots           & \vdots           & \vdots       & 1      & 1      & 0      &     0       &  \vdots   & \vdots         & \vdots          \\\hline
+   n-1      & x_{A,n-1}  &   z_{n-1} & \lambda_{1,n-1}  & \lambda_{2,n-1}  & x_{P[m_n]}   & 1      & 0      & 0      &     0       &  \vdots   & \vdots         & \vdots          \\\hline
+    0'      & x'_{A,0}   &   z'_0    & \lambda'_{1,0}   & \lambda'_{2,0}   & x_{P[m'_1]}  & 1      & 1      & 0      &     0       &  \vdots   & \vdots         & \vdots          \\\hline
+    1'      & x'_{A,1}   &   z'_1    & \lambda'_{1,1}   & \lambda'_{2,1}   & x_{P[m'_2]}  & 1      & 1      & 0      &     0       &  \vdots   & \vdots         & \vdots          \\\hline
+    2'      & x'_{A,2}   &   z'_2    & \lambda'_{1,2}   & \lambda'_{2,2}   & x_{P[m'_3]}  & 1      & 1      & 0      &     0       &  \vdots   & \vdots         & \vdots          \\\hline
+  \vdots    & \vdots     &   \vdots  & \vdots           & \vdots           & \vdots       & 1      & 1      & 0      &     0       &  \vdots   & \vdots         & \vdots          \\\hline
+   n-1'     & x'_{A,n-1} &   z'_{n-1}& \lambda'_{1,n-1} & \lambda'_{2,n-1} & x_{P[m'_n]}  & 1      & 2      & 2      &     0       &  \vdots   & \vdots         & \vdots          \\\hline
+    n'      & x'_{A,n}   &   0       & y_{A,n}          &                  &              & 0      & 0      & 0      &     0       &  \vdots   & \vdots         & \vdots          \\\hline
+\end{array}
+$$
+
+### Optimized Sinsemilla gate
+
+$\begin{array}{lrcl}
+\text{For } i \in [0, n), \text{ let} &x_{R,i} &=& \lambda_{1,i}^2 - x_{A,i} - x_{P,i} \\
+                                      &Y_{A,i} &=& (\lambda_{1,i} + \lambda_{2,i}) \cdot (x_{A,i} - x_{R,i}) \\
+                                      &y_{P,i} &=& Y_{A,i}/2 - \lambda_{1,i} \cdot (x_{A,i} - x_{P,i}) \\
+                                      &m_{i+1} &=& z_{i} - 2^k \cdot q_{S2,i} \cdot z_{i+1} \\
+                                      &q_{S3}  &=& q_{S2} \cdot (q_{S2} - 1)
+\end{array}$
+
+The Halo 2 circuit API can automatically substitute $y_{P,i}$, $x_{R,i}$, $y_{A,i}$, and $y_{A,i+1}$, so we don't need to do that manually.
+
+- $x_{A,0} = x_Q$
+- $2 \cdot y_Q = Y_{A,0}$
+- for $i$ from $0$ up to $n-1$:
+  - $(m_{i+1},\, x_{P,i},\, y_{P,i}) \in \mathcal{P}$
+  - $\lambda_{2,i}^2 = x_{A,i+1} + x_{R,i} + x_{A,i}$
+  - $4 \cdot \lambda_{2,i} \cdot (x_{A,i} - x_{A,i+1}) = 2 \cdot Y_{A,i} + (2 - q_{S3}) \cdot Y_{A,i+1} + 2 q_{S3} \cdot y_{A,n}$
+
+Note that each term of the last constraint is multiplied by $4$ relative to the constraint program given earlier. This is a small optimization that avoids divisions by $2$.
+
+$$
+\begin{array}{|c|l|}
+\hline
+\text{Degree} & \text{Constraint} \\\hline
+4   & fixed\rule{0.4em}{0.02ex}y_Q \cdot (2 \cdot fixed\rule{0.4em}{0.02ex}y_Q - Y_{A,0}) = 0 \\\hline
```

Comment:
For some reason the `_` in `fixed_y` causes a rendering error in mdbook, whether escaped as `\_` or not; hence the `\rule` workaround.